File: leetcode/python/131_palindrome_partitioning.py
class Solution:
    def __init__(self) -> None:
        self.cache: dict[str,str] = {}

    def partition(self, s: str) -> list[list[str]]:
        res: list[list[int]] = []
        if s in self.cache:
            return self.cache[s]
        for i in range(1, len(s) + 1):
            substr = s[:i]
            if not self.isPalindrome(substr):
                continue
                
            if combinations := self.partition(s[i:]):
                for item in combinations:
                    res.append([substr] + item)
            else:
                res.append([substr])
        
        self.cache[s] = res
        return res
    
    def isPalindrome(self, s: str) -> bool:
        for i in range(len(s) // 2):
            if s[i] != s[len(s)-1-i]:
                return False
        return True

if __name__ == '__main__':
    s = Solution()
    print(s.partition("aab"))
    print(s.partition("a"))
    print(s.partition("abbacb"))

File: leetcode/c++/131_palindrome_partitioning.cpp
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> out;
        partitionDFS(s, 0, out, res);
        return res;
    }

    void partitionDFS(string s, int start, vector<string> &out, vector<vector<string>> &res) {
        if (start == s.size()) {
            res.push_back(out);
            return;
        }
        for (int i = start; i < s.size(); i++) {
            if (isPalindrome(s, start, i)) {
                out.push_back(s.substr(start, i - start + 1));
                partitionDFS(s, i + 1, out, res);
                out.pop_back();
            }
        }
    }

    bool isPalindrome(string s, int low, int high) {
        while (low < high) {
            if (s[low++] != s[high--]) {
                return false;
            }
        }
        return true;
    }
};